### PR TITLE
Access Log: Move methods from FactoryContext to CommonFactoryContext

### PR DESCRIPTION
**Pull request by yanavlasov**
Additional Description:
Recent refactor in PR #15461 caused a number of build breakages in proprietary access log extensions. The CommonFactoryContext was missing some methods that these extensions used.

This is temporary patch to unblock broken extensions. Proper follow-up is tracked in issue #15864

I think going forward we may need to have two overloads of `createAccessLogInstance` one for access logging and another for logging proxied requests. It is possible that more and more methods move from FactoryContext to CommonFactoryContext.

Risk Level: Low, interface refactor.
Testing: Unit, integration tests.
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

**Review by davinci26 — COMMENTED** [review]
Apologies for breaking your internal build.

The approach SG to me although I do see one concern going forward. 

Since the `admin` interface / `server.h` is not a listener, I would assume it can not implement `listenerMetadata` and `listenerScope`. This means that the access loggers that you have developed internally would not work for the admin interface and they would cause an exception at initialisation.

What I am trying to say is that going forward with this path we are creating a 'gotcha' regarding which access loggers can be used by the admin interface.

**Comment by mattklein123** [comment]
> Since the admin interface / server.h is not a listener, I would assume it can not implement listenerMetadata and listenerScope. This means that the access loggers that you have developed internally would not work for the admin interface and they would cause an exception at initialisation.

The admin is actually listener so I don't think that's an issue. The bigger issue is that some of these methods are now getting added to the upstream context, which is decidedly not a listener. I think that is a bigger issue. :) In a perfect world, access loggers should work well for upstream logging as well as listener logging, so I think we are heading in the wrong direction here, and would likely have to split out the interfaces again in the future. Given this broke a build, I'm OK leaving some hacks in place, but can we potentially add some more TODOs to talk about how we want to handle this in the future?

**Comment by yanavlasov** [comment]
I've created an issue and added TODOs to the changes. I'm not quite sure what the solution should be. Should we add another interface between CommonFactoryContext -> ListenerFactoryContext -> FactoryContext ?

**Comment by mattklein123** [comment]
> I've created an issue and added TODOs to the changes. I'm not quite sure what the solution should be. Should we add another interface between CommonFactoryContext -> ListenerFactoryContext -> FactoryContext ?

Honestly I'm not sure. IMO the factory context interface structure has gotten quite messy at this point and probably needs to be holistically revisited. I think we probably need some other interface, yes, called something like ListenerAccessLogContext or something?

File: source/server/server.h [hunk 1]
```diff
@@ -208,6 +208,11 @@ class ServerFactoryContextImpl : public Configuration::ServerFactoryContext,
                ? server_.messageValidationContext().dynamicValidationVisitor()
                : server_.messageValidationContext().staticValidationVisitor();
   }
+  ProcessContextOptRef processContext() override { return server_.processContext(); }
+  const envoy::config::core::v3::Metadata& listenerMetadata() const override {
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+  }
+  Stats::Scope& listenerScope() override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
```
Comment (mattklein123) [inline source/server/server.h:215]: Can we get a similar TODO here?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
This change makes me pretty uncomfortable. I'm OK merging to unblock you, but are you able to follow up with a better fix later which is probably some new interface for access logs?

/wait

**Pull request closed without merging**